uk.ac.kcl.inf.modelling.xdsml.gemoc_henshin/src/uk/ac/kcl/inf/modelling/xdsml/gemoc_henshin/engine/solvers/HenshinSolver.xtend

```diff
@@ -0,0 +1,400 @@
+			if (!operator.parameters.reject[parameter|parameter.kind.equals(ParameterKind.VAR)].empty) {
+				println("Invalid unit with non-var parameters: " + operator.name)
+				return false
+			}
+		}
+
+		true
+	}
+	/**
+	 * configure the henshin solver setup, check if concurrent steps possible by running henshin CPA
+	 * and generate deleted nodes and edges if the feature enabled
+	 * @param henshin egraph, henshin engine and all semantic rules
+	 */
+	def configure(EGraph modelGraph, Engine henshinEngine, List<Rule> semanticRules){
+		this.modelGraph = modelGraph
+		this.henshinEngine = henshinEngine
+		var applicableRules = semanticRules.filter[r|r.checkParameters].toList
+		this.semanticRules = applicableRules
+		
+		//run Henshin CPA analysis
+		var cpa = new CpaByAGG();
+		var cpaOptions = new CPAOptions();
+		cpa.init(semanticRules, cpaOptions);
+		var result = cpa.runConflictAnalysis();
+		conflictPairs = result.getCriticalPairs();
+		
+		//get a list of conflict rules
+		conflictRules = new HashSet<Rule>();
+		for(var i = 0; i < conflictPairs.length; i++){
+			var cp = conflictPairs.get(i)
+			var currConflictKind = (cp as Conflict).getConflictKind();
+			//check that delete use is the only conflict kind present
+			//otherwise disable concurrent steps to avoid incorrect results
+			if(!currConflictKind.equals(ConflictKind.DELETE_USE_CONFLICT)){
+				this.showConcurrentSteps = false;
+				i = conflictPairs.length
+			}
+			var first = cp.getFirstRule();
+			var second = cp.getSecondRule();
+			conflictRules.add(first);
+			conflictRules.add(second);
+		}
+		
+		//if the feature enabled
+		//get deleted nodes and edges for each rule and store
+		if(showConcurrentSteps){
+			ruleDeletedNodesMap = new HashMap<Rule, ArrayList<Node>>();
+			ruleDeletedEdgesMap = new HashMap<Rule, ArrayList<Edge>>();
+			for(Rule r: conflictRules){
+				var deletedNodes = getDeletedNodes(r);
+				var deletedEdges = getDeletedEdges(r);
+				ruleDeletedNodesMap.put(r, deletedNodes);
+				ruleDeletedEdgesMap.put(r, deletedEdges);
+			}	
+		}	
+	}
+	
+	/**
+	 * get deleted nodes from a rule
+	 * @param a rule
+	 */
+	def getDeletedNodes(Rule rule) {
+		var lhsNodes = rule.getLhs().getNodes();
+		var rhs = rule.getRhs();
+		var mappings = rule.getMappings();
+		var deletedNodes =  new ArrayList<Node>();
+		
+		//check if the LHS nodes are mapped to something in RHS
+		for(Node eo: lhsNodes){
+			if(mappings.getImage(eo,rhs) === null)
+				deletedNodes.add(eo);
+		}
+		deletedNodes;
+	}
+	/**
+	 * get deleted edges from a rule
+	 * @param a rule
+	 */
+	def getDeletedEdges(Rule rule) {
+		var rhs = rule.getRhs();
+		var mappings = rule.getMappings();
+		var deletedEdges =  new ArrayList<Edge>();
+		var lhsEdges = rule.getLhs().getEdges();
+		
+		//check if the LHS edges are mapped to something in RHS
+		for(Edge e: lhsEdges){
+			if(mappings.getImage(e,rhs) === null)
+				deletedEdges.add(e);
+		}
+		deletedEdges;
+	}
+	
+	override getState() {
```

Comment:
All of these methods that we don't actually need should get a brief bit of documentation explaining why they're here. Again, it would be good to add an official issue to the relevant GEMOC repository so that we can track their progress on refactoring this.